### PR TITLE
Clearnet p2p

### PR DESCRIPTION
**Pull request by sachaaaaa**
- generate certs and stuff for SSL, after storage is ready
- spin an HTTPS server for p2p instead of HTTP
- use `nat-upnp` to punch hole. We try to read the existing mappings first, if that fails, we assume uPnP not enabled and we stop the server. If that works, we get the existing public ports in use. Then we try to punch from port 22100 to 22200, skipping the ports we know are already in use by other processes. After this, we can query `nat-upnp` for the public ip that we need to broadcast.
- Allowed half p2p (one client with p2p, the other not) by adding a type REACHABLE/UNREACHABLE sent in the broadcast message. UNREACHABLE means that the client can't spin its HTTPS server but still needs to send that message to trigger the code path that allows half p2p.

Coming next: signing with signal pubkey of SSL cert + verification

### CODE REVIEW

File: libloki/modules/local_loki_server.js [hunk 1]
```diff
@@ -79,18 +93,90 @@ class LocalLokiServer extends EventEmitter {
...
+            resolve();
+        });
+      });
+      try {
+        // eslint-disable-next-line no-await-in-loop
+        await p;
+        this.publicPort = publicPort;
+        this.timerHandler = setTimeout(() => {
+          try {
+            this.publicPort = this.punchHole();
```
Comment (Mikunj) [inline libloki/modules/local_loki_server.js]: Does there need to be an `await` here?

Comment (sachaaaaa) [reply libloki/modules/local_loki_server.js]: Excellent catch...

File: preload.js [hunk 2]
```diff
@@ -50,6 +51,19 @@ window.isBeforeVersion = (toCheck, baseVersion) => {
 
 // temporary clearnet fix
 process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
+window.getSelfSignedCert = () => {
+  let pems = window.storage.get('self-signed-certificate', null);
+  if (!pems) {
+    const pubKey = window.storage.get('number_id');
+    const attrs = [{ name: 'commonName', value: pubKey }];
+    pems = selfsigned.generate(attrs, { days: 365 * 10 });
```
Comment (BeaudanBrown) [inline preload.js]: How long will this take to generate on average?

Comment (BeaudanBrown) [reply preload.js]: Also you are awaiting this function above, does selfsigned.generate return a promise?

File: libloki/modules/local_loki_server.js [hunk 3]
```diff
@@ -14,9 +16,21 @@ class LocalLokiServer extends EventEmitter {
...
-  constructor() {
+  constructor(pems, options) {
     super();
-    this.server = http.createServer((req, res) => {
+    const httpsOptions = {
+      key: pems.private,
+      cert: pems.cert,
+    };
+    // eslint-disable-next-line no-param-reassign
+    options = {
```
[the remark this answers is not in the document]
Comment (sachaaaaa) [reply libloki/modules/local_loki_server.js]: Yeah, what was I thinking...

File: libloki/modules/local_loki_server.js [hunk 4]
```diff
@@ -79,18 +93,90 @@ class LocalLokiServer extends EventEmitter {
...
-      this.server.listen(port, ip, err => {
+      this.server.listen(port, ip, async (err) => {
         if (err) {
           rej(err);
+        } else if (this.upnpClient) {
+          try {
+            const publicPort = await this.punchHole();
+            res(publicPort);
+          } catch (e) {
+            if (e instanceof textsecure.HolePunchingError)
```
[the remark this answers is not in the document]
Comment (sachaaaaa) [reply libloki/modules/local_loki_server.js]: Absolutely right

File: preload.js (hunk 2 above)
Comment (sachaaaaa) [reply preload.js]: `selfsigned.generate` is a sync function, maybe you got confused with `await window.getSelfSignedCert();`?

A quick test in node gives 150-200 ms per call.

Comment (BeaudanBrown) [reply preload.js]: Sorry that's what I meant, you are awaiting `window.getSelfSignedCert()` but it isn't async so I thought maybe the generate function was returning a promise

You probably don't want to await `window.getSelfSignedCert()` then?

File: libloki/modules/local_loki_server.js [hunk 5]
```diff
@@ -79,18 +88,91 @@ class LocalLokiServer extends EventEmitter {
...
+          // to allow reusing that port.
+          resolve(results
+            .filter(entry => entry.private.port !== privatePort)
+            .map(entry => entry.public.port));
+        }
+      });
+    });
+
+    for (let publicPort = portStart; publicPort <= portEnd; publicPort += 1) {
+      if (publicPortsInUse.includes(publicPort))
```
Comment (BeaudanBrown) [inline libloki/modules/local_loki_server.js]: There are a few other if/else without curly's if we are going to enforce it...

File: preload.js (hunk 2 above)
Comment (sachaaaaa) [reply preload.js]: Oh I see, you're right. Removed unnecessary awaits.

File: libloki/modules/local_loki_server.js (hunk 5 above)
Comment (sachaaaaa) [reply libloki/modules/local_loki_server.js]: I wish the linter would take care of that ...
Anyway, fixed!

Comment (BeaudanBrown) [reply libloki/modules/local_loki_server.js]: Ya we should see if there is a setting we can add to eslint...
One day™...

File: js/background.js [hunk 6]
```diff
@@ -535,9 +548,11 @@
     window.log.info('Cleanup: complete');
 
     window.log.info('listening for registration events');
-    Whisper.events.on('registration_done', () => {
+    Whisper.events.on('registration_done', async () => {
```
Comment (Mikunj) [inline js/background.js]:
```suggestion
    Whisper.events.on('registration_done', () => {
```
Un-needed async

Comment (BeaudanBrown) [reply js/background.js]: Will do in separate PR

**Pull request merged**